runtime/storage/cache/cache.go

```diff
@@ -0,0 +1,267 @@
+
+const (
+	// AllKeysLRU keeps most recently used keys and removes least recently used (LRU) keys.
+	// This is a good default choice for most cache use cases if you're not sure.
+	AllKeysLRU EvictionPolicy = "allkeys-lru"
+
+	// AllKeysLFU keeps frequently used keys and removes least frequently used (LFU) keys.
+	AllKeysLFU EvictionPolicy = "allkeys-lfu"
+
+	// AllKeysRandom randomly removes keys as needed.
+	AllKeysRandom EvictionPolicy = "allkeys-random"
+
+	// VolatileLRU removes least recently used keys with an expiration set.
+	// It behaves like NoEviction if there are no keys to evict with an expiration set.
+	VolatileLRU EvictionPolicy = "volatile-lru"
+
+	// VolatileLFU removes least frequently used keys with an expiration set.
+	// It behaves like NoEviction if there are no keys to evict with an expiration set.
+	VolatileLFU EvictionPolicy = "volatile-lfu"
+
+	// VolatileTTL removes keys with an expiration set and the shortest time to live (TTL).
+	// It behaves like NoEviction if there are no keys to evict with an expiration set.
+	VolatileTTL EvictionPolicy = "volatile-ttl"
+
+	// VolatileRandom randomly removes keys with an expiration set.
+	// It behaves like NoEviction if there are no keys to evict with an expiration set.
+	VolatileRandom EvictionPolicy = "volatile-random"
+
+	// NoEviction does not evict any keys, and instead returns an error to the client
+	// when the max memory limit is reached.
+	NoEviction EvictionPolicy = "noeviction"
+)
+
+//publicapigen:keep
+type constStr string
+
+// Cluster represents a Redis cache cluster.
+type Cluster struct {
+	cfg ClusterConfig
+	mgr *Manager
+	cl  *redis.Client
+}
+
+// KeyspaceConfig specifies the configuration options for a cache keyspace.
+type KeyspaceConfig struct {
+	// KeyPattern is a string literal representing the
+	// cache key pattern for this keyspace.
+	KeyPattern constStr
+
+	// DefaultExpiry specifies the default key expiry for cache items
+	// in this keyspace.
+	//
+	// When set, all write operations set (for new keys)
+	// or update (for existing keys) the expiration time.
+	//
+	// When updating the expiration time Encore always
+	// performs the combined operation atomically.
+	//
+	// If nil, cache items have no expiry date by default.
+	//
+	// The default behavior can be overridden by passing in
+	// an ExpiryFunc or KeepTTL as a WriteOption to a specific operation.
+	DefaultExpiry ExpiryFunc
+
+	// EncoreInternal_DefLoc specifies where the keyspace is defined.
+	// It's an internal field set by Encore's compiler.
+	//publicapigen:drop
+	EncoreInternal_DefLoc int32
+
+	// EncoreInternal_KeyMapper specifies how typed keys are translated
+	// to a string. It's of type any to avoid making KeyspaceConfig
+	// a generic type. It's an internal field set by Encore's compiler.
+	//
+	// The type must be func(K) string.
+	//publicapigen:drop
+	EncoreInternal_KeyMapper any
+}
+
+// An OpError describes the operation that failed.
+type OpError struct {
```

Comment:
Why isn't this an `errs` type? It feels very weird that we have our own error type, and yet we have here another type of error again.

Reply:
They serve very different use cases. `encore.dev/beta/errs` is for application-returned errors and API responses. This type is for caching-specific information like the key(s) and cache operation. Much like the standard library has lots of different error types.

I don't see why we necessarily need to have exactly one error type given that Encore supports lots of different use cases. (We should probably have a separate error type for `sqldb` errors, too.)